packages/flutter/lib/src/semantics/binding.dart

```diff
@@ -28,12 +33,124 @@ mixin SemanticsBinding on BindingBase {
   static SemanticsBinding get instance => BindingBase.checkInstance(_instance);
   static SemanticsBinding? _instance;
 
+  /// Whether semantics information must be collected.
+  ///
+  /// Returns true if either the platform has requested semantics information
+  /// to be generated or if [ensureSemantics] has been called otherwise.
+  ///
+  /// To get notified when this value changes register a listener with
+  /// [addSemanticsEnabledListener].
+  bool get semanticsEnabled {
+    assert(_semanticsEnabled.value == (_outstandingHandles > 0));
+    return _semanticsEnabled.value;
+  }
+  late final ValueNotifier<bool> _semanticsEnabled = ValueNotifier<bool>(platformDispatcher.semanticsEnabled);
+
+  /// Adds a `listener` to be called when [semanticsEnabled] changes.
+  ///
+  /// See also:
+  ///
+  ///  * [removeSemanticsEnabledListener] to remove the listener again.
+  ///  * [ValueNotifier.addListener], which documents how and when listeners are
+  ///    called.
+  void addSemanticsEnabledListener(VoidCallback listener) {
+    _semanticsEnabled.addListener(listener);
+  }
+
+  /// Removes a `listener` added by [addSemanticsEnabledListener].
+  ///
+  /// See also:
+  ///
+  ///  * [ValueNotifier.removeListener], which documents how listeners are
+  ///    removed.
+  void removeSemanticsEnabledListener(VoidCallback listener) {
+    _semanticsEnabled.removeListener(listener);
+  }
+
+  int _outstandingHandles = 0;
+
+  /// Creates a new [SemanticsHandle] and requests the collection of semantics
+  /// information.
+  ///
+  /// Semantics information are only collected when there are clients interested
+  /// in them. These clients express their interest by holding a
+  /// [SemanticsHandle].
+  ///
+  /// Clients can close their [SemanticsHandle] by calling
+  /// [SemanticsHandle.dispose]. Once all outstanding [SemanticsHandle] objects
+  /// are closed, semantics information are no longer collected.
+  SemanticsHandle ensureSemantics() {
+    assert(_outstandingHandles >= 0);
+    _outstandingHandles++;
+    assert(_outstandingHandles > 0);
+    _semanticsEnabled.value = true;
+    return SemanticsHandle._(_didDisposeSemanticsHandle);
+  }
+
+  void _didDisposeSemanticsHandle() {
+    assert(_outstandingHandles > 0);
+    _outstandingHandles--;
+    assert(_outstandingHandles >= 0);
+    _semanticsEnabled.value = _outstandingHandles > 0;
+  }
+
+  // Handle for semantics request from the platform.
+  SemanticsHandle? _semanticsHandle;
+
+  void _handleSemanticsEnabledChanged() {
+    if (platformDispatcher.semanticsEnabled) {
+      _semanticsHandle ??= ensureSemantics();
+    } else {
+      _semanticsHandle?.dispose();
+      _semanticsHandle = null;
+    }
+  }
+
+  void _handleSemanticsAction(int id, ui.SemanticsAction action, ByteData? args) {
+    performSemanticsAction(SemanticsActionEvent(
```

Comment:
This indirection doesn't seem to add too much value.. maybe just leave _handleSemanticsAction public and let RenderingBinding to handle raw args directly?

Reply:
I do believe it is the responsibility of the semantics layer to decode args as it is the one that picks the messageCodec. If we were to decide that we need to change the message codec used, we should do the change in one place here instead of fixing up client code elsewhere.

Furthermore, the SemanticsActionEvent gives us a little more flexibility in the future to extend the information associated with a SemanticsEvent without breaking the method signature. For example, we may want to add the view ID from where the event was originating in the future.